File: Weekly_Modelling_Python/Script/2_weekly_model_predictions.py
"""
2. Weekly Model Predictions (Python)
Mirrors: Weekly_Modelling/Script/2. Weekly Model Predictions.R

Loads trained model packages and generates Win / Top5 / Top10 / Top20 predictions
for the upcoming tournament field.

Output columns match R output for direct comparison:
  Surname, Firstname, Rating, Market_Odds, Model_Score,
  Probability, Normalised_Probability, Normalised_Model_Odds

Run: python 2_weekly_model_predictions.py
"""

import sys
import warnings
from datetime import datetime
from pathlib import Path

import joblib
import numpy as np
import pandas as pd

try:
    sys.path.insert(0, str(Path(__file__).parent))
except NameError:  # IPython / Jupyter — assume CWD is the Script dir
    sys.path.insert(0, str(Path.cwd()))
from config import (
    BASE_MODEL_VARS,
    BETTING_MARKETS,
    MODELS_DIR,
    PREDICTIONS_DIR,
    RANDOM_SEED,
    SEASON_SUFFIX,
    TOUR_CONFIG,
)

warnings.filterwarnings("ignore")


# ===== HELPERS =====

def get_market_vars(market_config: dict) -> list:
    odds_cols = {"Win_odds", "Top5_odds", "Top10_odds", "Top20_odds"}
    return [v for v in BASE_MODEL_VARS if v not in odds_cols]


def ensemble_predict(market_pkg: dict, X: np.ndarray, odds: np.ndarray) -> np.ndarray:
    """
    1. Get predictions from each base model.
    2. Stack with implied odds as meta-features.
    3. Apply meta-model to produce calibrated probabilities.
    """
    model_preds = np.column_stack([
        model.predict_proba(X)[:, 1]
        for model in market_pkg["models"].values()
    ])
    meta_X_scaled = market_pkg["meta_scaler"].transform(model_preds)
    return (
        market_pkg["meta_model"].predict_proba(meta_X_scaled)[:, 1],
        model_preds.mean(axis=1),   # raw model score (mean of individual model probs)
    )


# ===== PREDICTION FOR ONE MARKET =====

def predict_market(market_name: str, market_pkg: dict, newdat: pd.DataFrame) -> pd.DataFrame:
    market_config = BETTING_MARKETS[market_name]
    model_vars    = market_pkg["model_vars"]
    odds_col      = market_pkg["odds_col"]

    available = [v for v in model_vars if v in newdat.columns]
    missing   = [v for v in model_vars if v not in newdat.columns]
    if missing:
        print(f"    Warning: missing vars (skipped): {missing}")

    df = newdat[available + [odds_col]].copy()
    df = df.dropna()
    X    = df[available].values.astype(float)
    odds = df[odds_col].values.astype(float)

    proba, raw_score = ensemble_predict(market_pkg, X, odds)

    # Tournament-level normalisation: probabilities sum to market_size
    market_size = market_pkg["market_size"]
    prob_sum    = proba.sum()
    norm_prob   = (proba / prob_sum) * market_size if prob_sum > 0 else proba

    results = newdat.loc[df.index].copy()
    results["Model_Score"]            = raw_score.round(5)
    results["Probability"]            = proba.round(6)
    results["Normalised_Probability"] = norm_prob.round(6)
    results["Model_Odds"]             = (1.0 / proba.clip(1e-8)).round(2)
    results["Normalised_Model_Odds"]  = (1.0 / norm_prob.clip(1e-8)).round(2)

    # Sort ascending by market odds
    results = results.sort_values(odds_col)
    return results


# ===== EXPORT =====

def export_predictions(tour_results: dict, tour_key: str) -> Path:
    PREDICTIONS_DIR.mkdir(parents=True, exist_ok=True)
    date_str = datetime.now().strftime("%d-%m-%Y")
    out_path = PREDICTIONS_DIR / f"{tour_key}_Predictions_{date_str}.xlsx"

    with pd.ExcelWriter(out_path, engine="openpyxl") as writer:
        for market_name, df in tour_results.items():
            if df is None or len(df) == 0:
                continue

            odds_col = BETTING_MARKETS[market_name]["odds_col"]

            # Select and rename columns to match R output
            cols_available = []
            rename_map = {}
            for col in ["surname", "firstname"]:
                if col in df.columns:
                    cols_available.append(col)
                    rename_map[col] = col.capitalize()

            out_df = pd.DataFrame()
            out_df["Surname"]   = df.get("surname",   df.get("Surname",   ""))
            out_df["Firstname"] = df.get("firstname", df.get("Firstname", ""))
            if "rating" in df.columns:
                out_df["Rating"] = df["rating"]
            out_df["Market_Odds"]           = df[odds_col]
            out_df["Model_Score"]           = df["Model_Score"]
            out_df["Probability"]           = df["Probability"]
            out_df["Normalised_Probability"] = df["Normalised_Probability"]
            out_df["Normalised_Model_Odds"] = df["Normalised_Model_Odds"]

            out_df.to_excel(writer, sheet_name=f"{market_name}_Market", index=False)

    print(f"  Predictions saved: {out_path.name}")
    return out_path


# ===== PROCESS ONE TOUR =====

def process_tour(tour_key: str, tour_info: dict):
    print(f"\n{'='*60}")
    print(f"PREDICTING: {tour_info['name']}")
    print(f"{'='*60}")

    model_path = MODELS_DIR / f"{tour_key}_Trained_Models_{SEASON_SUFFIX}.pkl"
    if not model_path.exists():
        print(f"  Model file not found: {model_path}")
        return None

    weekly_path = tour_info["weekly_file"]
    if not weekly_path.exists():
        print(f"  Weekly file not found: {weekly_path}")
        return None

    package = joblib.load(model_path)
    print(f"  Model trained: {package['trained_at'].strftime('%Y-%m-%d %H:%M')}")

    newdat = pd.read_excel(weekly_path).dropna()
    print(f"  Field size: {len(newdat)} players")

    market_results = {}
    for market_name, market_pkg in package["markets"].items():
        print(f"  Processing {market_name}...")
        df = predict_market(market_name, market_pkg, newdat)
        market_results[market_name] = df
        print(f"    {len(df)} players  |  "
              f"prob range: {df['Probability'].min():.4f}–{df['Probability'].max():.4f}")

    return market_results


# ===== MAIN =====

def main():
    print("=== GOLF TOURNAMENT PREDICTIONS (PYTHON) ===")
    print(f"Season: {SEASON_SUFFIX}  |  Date: {datetime.now().strftime('%d-%m-%Y')}")

    all_results = {}
    output_files = {}

    for tour_key, tour_info in TOUR_CONFIG.items():
        try:
            results = process_tour(tour_key, tour_info)
            if results:
                all_results[tour_key] = results
                output_files[tour_key] = export_predictions(results, tour_key)
        except Exception as e:
            print(f"\nError processing {tour_key}: {e}")
            import traceback
            traceback.print_exc()

    print("\n=== PREDICTION SUMMARY ===")
    for tour_key, results in all_results.items():
        print(f"\n{TOUR_CONFIG[tour_key]['name']}:")
        for market_name, df in results.items():
            if df is not None:
                print(f"  {market_name}: {len(df)} players")
        if tour_key in output_files:
            print(f"  Output: {output_files[tour_key].name}")

    if not all_results:
        print("\nNo predictions generated. Check that:")
        print("  1. Processed weekly files exist in Weekly_Modelling_Python/Input/")
        print("  2. Trained models exist in Weekly_Modelling_Python/Output/Models/")
        print("  3. Run 1_weekly_data_preprocessing.py and seasonal_model_training.py first")

    print("\n=== COMPLETE ===")


if __name__ == "__main__":
    main()

File: Weekly_Modelling/Script/2. Weekly Model Predictions.R
# ===== CONFIGURATION =====
library(biomod2)
library(dplyr)
library(readxl)
library(openxlsx)
library(lubridate)

# Set working directory
setwd("C:/Projects/Golf/Weekly_Modelling")

# Define tour configurations
TOUR_CONFIG <- list(
  PGA = list(
    name = "PGA Tour",
    models_file = "./Output/Models/PGA_Trained_Models_S26.rds",
    upcoming_file = "./Input/This_Week_PGA_Processed.xlsx"
  ),
  Euro = list(
    name = "European Tour", 
    models_file = "./Output/Models/Euro_Trained_Models_S26.rds",
    upcoming_file = "./Input/This_Week_Euro_Processed.xlsx"
  )
)

# Define base model variables (same as training script)
BASE_MODEL_VARS <- c("rating_vs_field_best",
                     "rating",
                     "rating_vs_field_worst", 
                     "yr3_All",
                     "Top5_rank",
                     "compat2",
                     "sgp_field_zscore",
                     "sgtee_field_zscore",
                     "course",
                     "sgatg_vs_field_median", 
                     "sgapp_vs_field_median",
                     "Starts_Not10",
                     "current",
                     "location", 
                     "field")

# Define betting markets configuration.
# Odds columns are intentionally excluded from all model layers —
# empirical testing showed worse outcomes when implied probability
# was included as a model feature.
BETTING_MARKETS <- list(
  "Winner" = list(target_col = "win",   odds_col = "Win_odds",  market_size = 1),
  "Top5"   = list(target_col = "top_5", odds_col = "Top5_odds", market_size = 5),
  "Top10"  = list(target_col = "top_10", odds_col = "Top10_odds", market_size = 10),
  "Top20"  = list(target_col = "top_20", odds_col = "Top20_odds", market_size = 20)
)

# ===== HELPER FUNCTIONS =====

# Apply Platt scaling to new scores using fitted parameters
predict_platt_scaling <- function(scores, platt_params) {
  1 / (1 + exp(platt_params$A * scores + platt_params$B))
}

# Get market-specific model variables.
# Odds columns are excluded from all model layers — they are only used for
# EV/edge calculations at prediction time.
get_market_model_vars <- function(market_config, base_model_vars) {
  base_vars_no_odds <- base_model_vars[!base_model_vars %in% c("Win_odds", "Top5_odds", "Top10_odds", "Top20_odds")]
  return(base_vars_no_odds)
}

# ===== PREDICTION FUNCTIONS =====

# Function to make predictions for a single model/run
predict_single_run <- function(ensemble_model, calibration_models, newdat, market_config, 
                               market_name, run_number, tour_key) {
  
  cat("    Processing Run", run_number, "\n")
  
  # Get market-specific variables
  market_model_vars <- get_market_model_vars(market_config, BASE_MODEL_VARS)
  
  # Check available variables
  available_vars <- market_model_vars[market_model_vars %in% names(newdat)]
  
  if (length(available_vars) < length(market_model_vars)) {
    missing <- market_model_vars[!market_model_vars %in% names(newdat)]
    cat("      Warning: Missing variables:", paste(missing, collapse = ", "), "\n")
  }
  
  # Prepare prediction data
  pred_env <- newdat[, available_vars]
  pred_env_xy <- newdat[, c("surname", "firstname")]
  
  # Make ensemble predictions
  proj_name <- paste0(tour_key, "_", market_name, "_Run", run_number)
  
  biomod_proj <- BIOMOD_EnsembleForecasting(
    bm.em = ensemble_model,
    proj.name = proj_name,
    new.env = pred_env,
    new.env.xy = pred_env_xy,
    models.chosen = "all",
    metric.binary = "TSS",
    metric.filter = "TSS",
    na.rm = TRUE
  )
  
  # Extract raw model scores
  model_scores <- biomod_proj@proj.out@val$pred
  
  # Create results dataframe
  results <- data.frame(
    surname = pred_env_xy$surname,
    firstname = pred_env_xy$firstname,
    model_score = model_scores,
    market_odds = newdat[[market_config$odds_col]]
  )

  # Apply Platt scaling calibration (odds excluded from all model layers)
  results$final_probability <- predict_platt_scaling(model_scores, calibration_models$platt)

  return(results)
}

# Function to process all runs for a market and average results
process_market_predictions <- function(market_name, market_training_results, 
                                       newdat, tour_key) {
  
  cat("  Processing", market_name, "market\n")
  
  market_config <- BETTING_MARKETS[[market_name]]
  run_predictions <- list()
  
  # Process each run
  for (run_idx in 1:length(market_training_results$run_results)) {
    run_data <- market_training_results$run_results[[run_idx]]
    
    predictions <- predict_single_run(
      ensemble_model = run_data$ensemble_model,
      calibration_models = run_data$calibration_models,
      newdat = newdat,
      market_config = market_config,
      market_name = market_name,
      run_number = run_idx,
      tour_key = tour_key
    )
    
    run_predictions[[run_idx]] <- predictions
  }
  
  # Average predictions across runs
  cat("    Averaging", length(run_predictions), "runs\n")
  
  # Start with first run structure
  averaged_results <- run_predictions[[1]]
  
  if (length(run_predictions) > 1) {
    # Sum up scores and probabilities
    for (run_idx in 2:length(run_predictions)) {
      averaged_results$model_score <- averaged_results$model_score + 
        run_predictions[[run_idx]]$model_score
      averaged_results$final_probability <- averaged_results$final_probability + 
        run_predictions[[run_idx]]$final_probability
      
      # Also average the other calibration methods if present
      if ("prob_glm_odds" %in% names(averaged_results)) {
        averaged_results$prob_glm_odds <- averaged_results$prob_glm_odds + 
          run_predictions[[run_idx]]$prob_glm_odds
      }
      if ("prob_glm_prob" %in% names(averaged_results)) {
        averaged_results$prob_glm_prob <- averaged_results$prob_glm_prob + 
          run_predictions[[run_idx]]$prob_glm_prob
      }
      if ("prob_platt" %in% names(averaged_results)) {
        averaged_results$prob_platt <- averaged_results$prob_platt + 
          run_predictions[[run_idx]]$prob_platt
      }
    }
    
    # Divide by number of runs to get average
    n_runs <- length(run_predictions)
    averaged_results$model_score <- averaged_results$model_score / n_runs
    averaged_results$final_probability <- averaged_results$final_probability / n_runs
    
    if ("prob_glm_odds" %in% names(averaged_results)) {
      averaged_results$prob_glm_odds <- averaged_results$prob_glm_odds / n_runs
    }
    if ("prob_glm_prob" %in% names(averaged_results)) {
      averaged_results$prob_glm_prob <- averaged_results$prob_glm_prob / n_runs
    }
    if ("prob_platt" %in% names(averaged_results)) {
      averaged_results$prob_platt <- averaged_results$prob_platt / n_runs
    }
  }
  
  # Normalise probabilities to sum to market size
  market_size <- market_config$market_size
  total_prob <- sum(averaged_results$final_probability)
  averaged_results$normalised_probability <- (averaged_results$final_probability / total_prob) * market_size
  
  # Calculate model odds
  averaged_results$model_odds <- round(1 / averaged_results$final_probability, 2)
  averaged_results$normalised_model_odds <- round(1 / averaged_results$normalised_probability, 2)
  
  # Add rating if available
  if ("rating" %in% names(newdat)) {
    averaged_results$rating <- newdat$rating[match(paste(averaged_results$surname, averaged_results$firstname),
                                                   paste(newdat$surname, newdat$firstname))]
  }
  
  # Sort by market odds
  averaged_results <- averaged_results[order(averaged_results$market_odds), ]
  
  return(averaged_results)
}

# Function to process all markets for a tour
process_tour_predictions <- function(tour_key, tour_config) {
  
  cat("\n=== PROCESSING", toupper(tour_config$name), "===\n")
  
  # Load trained models
  if (!file.exists(tour_config$models_file)) {
    cat("  Error: Models file not found:", tour_config$models_file, "\n")
    return(NULL)
  }
  
  cat("  Loading models from:", tour_config$models_file, "\n")
  trained_models <- readRDS(tour_config$models_file)
  
  # Load upcoming event data
  if (!file.exists(tour_config$upcoming_file)) {
    cat("  Error: Upcoming event file not found:", tour_config$upcoming_file, "\n")
    return(NULL)
  }
  
  cat("  Loading upcoming event from:", tour_config$upcoming_file, "\n")
  newdat <- read_excel(tour_config$upcoming_file)
  newdat <- newdat[complete.cases(newdat), ]
  cat("  Found", nrow(newdat), "players\n")
  
  # Process each market
  market_predictions <- list()
  
  for (market_name in names(trained_models)) {
    market_results <- process_market_predictions(
      market_name = market_name,
      market_training_results = trained_models[[market_name]],
      newdat = newdat,
      tour_key = tour_key
    )
    
    market_predictions[[market_name]] <- market_results
  }
  
  return(list(
    tour_name = tour_config$name,
    tour_key = tour_key,
    predictions = market_predictions,
    field_size = nrow(newdat)
  ))
}

# Function to export predictions to Excel
export_predictions <- function(tour_results, output_dir = "./Output/Predictions/") {
  
  if (is.null(tour_results)) {
    cat("  Warning: tour_results is NULL\n")
    return(NULL)
  }
  
  if (is.null(tour_results$predictions) || length(tour_results$predictions) == 0) {
    cat("  Warning: No predictions to export\n")
    return(NULL)
  }
  
  # Create output directory if it doesn't exist
  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }
  
  # Create workbook
  wb <- createWorkbook()
  
  sheets_added <- 0
  
  for (market_name in names(tour_results$predictions)) {
    market_data <- tour_results$predictions[[market_name]]
    
    if (is.null(market_data) || nrow(market_data) == 0) {
      cat("    Warning: No data for", market_name, "market\n")
      next
    }
    
    # Add individual market sheet
    sheet_name <- paste0(market_name, "_Market")
    addWorksheet(wb, sheet_name)
    
    # Prepare data for export with renamed and capitalised columns
    export_data <- data.frame(
      Surname = market_data$surname,
      Firstname = market_data$firstname,
      Rating = if(!is.null(market_data$rating)) market_data$rating else NA,
      Market_Odds = market_data$market_odds,
      Model_Score = market_data$model_score,
      Probability = market_data$final_probability,  # Renamed from final_probability
      Normalised_Probability = market_data$normalised_probability,  # British spelling
      Normalised_Model_Odds = market_data$normalised_model_odds  # British spelling
    )
    
    writeData(wb, sheet_name, export_data)
    sheets_added <- sheets_added + 1
  }
  
  if (sheets_added == 0) {
    cat("  Warning: No sheets were added to workbook\n")
    return(NULL)
  }
  
  # Save workbook
  output_file <- paste0(output_dir, tour_results$tour_key, "_Predictions_", 
                        format(Sys.time(), "%d-%m-%Y"), ".xlsx")
  
  tryCatch({
    saveWorkbook(wb, output_file, overwrite = TRUE)
    cat("  Predictions saved to:", output_file, "\n")
    return(output_file)
  }, error = function(e) {
    cat("  Error saving workbook:", e$message, "\n")
    return(NULL)
  })
}

# ===== MAIN EXECUTION =====

cat("=== GOLF TOURNAMENT PREDICTIONS FROM TRAINED MODELS ===\n")

# Store results for all tours
all_tour_results <- list()
output_files <- list()

# Process each tour
for (tour_key in names(TOUR_CONFIG)) {
  tour_config <- TOUR_CONFIG[[tour_key]]
  
  tryCatch({
    # Process predictions for this tour
    tour_results <- process_tour_predictions(tour_key, tour_config)
    
    if (!is.null(tour_results)) {
      all_tour_results[[tour_key]] <- tour_results
      
      # Export to Excel
      output_file <- export_predictions(tour_results)
      if (!is.null(output_file)) {
        output_files[[tour_key]] <- output_file
      }
    }
  }, error = function(e) {
    cat("\nError processing", tour_config$name, ":", e$message, "\n")
    cat("Continuing with other tours...\n")
  })
}

# ===== FINAL SUMMARY =====

cat("\n=== PREDICTION SUMMARY ===\n")

if (length(all_tour_results) > 0) {
  for (tour_key in names(all_tour_results)) {
    tour_results <- all_tour_results[[tour_key]]
    cat("\n", tour_results$tour_name, ":\n")
    cat("  Field size:", tour_results$field_size, "players\n")
    cat("  Markets processed:", paste(names(tour_results$predictions), collapse = ", "), "\n")
    if (!is.null(output_files[[tour_key]])) {
      cat("  Output file:", basename(output_files[[tour_key]]), "\n")
    } else {
      cat("  Output file: Failed to save\n")
    }
  }
  
} else {
  cat("\nNo predictions were generated. Please check:\n")
  cat("1. Model files exist in ./Output/Models/\n")
  cat("2. Upcoming event files exist in ./Input/\n")
  cat("3. File names match the configuration\n")
}

cat("=== COMPLETE ===\n")